tensorflow/compiler/mlir/tosa/tests/tfl-to-tosa-pipeline.mlir

```diff
@@ -1011,34 +1011,137 @@ func.func @test_max_pool2d_dynamic(%arg0: tensor<?x32x32x8xf32>) -> tensor<*xf32
   %cst = arith.constant dense<[1, 819]> : tensor<2xi32>
   %0 = "tfl.reshape"(%arg0, %cst) : (tensor<13x21x3xf32>, tensor<2xi32>) -> tensor<*xf32>
   func.return %0 : tensor<*xf32>
 }
 
 // -----
 
-// CHECK-LABEL: test_reshape_unknown
-// CHECK: %[[VAR0:.*]] = "tosa.reshape"(%arg0) <{new_shape = array<i64: 9, -1>}>
-// CHECK-SAME: -> tensor<9x91xf32>
-func.func @test_reshape_unknown(%arg0: tensor<13x21x3xf32>) -> tensor<*xf32> {
-  %cst = arith.constant dense<[9, -1]> : tensor<2xi32>
-  %0 = "tfl.reshape"(%arg0, %cst) : (tensor<13x21x3xf32>, tensor<2xi32>) -> tensor<*xf32>
-  func.return %0 : tensor<*xf32>
-}
-
-// -----
-
-// CHECK-LABEL: test_reshape_dynamic
-// CHECK: %[[VAR0:.*]] = "tosa.reshape"(%arg0) <{new_shape = array<i64: 3, -1>}>
-// CHECK-SAME: -> tensor<3x?xf32>
-func.func @test_reshape_dynamic(%arg0: tensor<13x21x?xf32>) -> tensor<*xf32> {
-  %cst = arith.constant dense<[3, -1]> : tensor<2xi32>
-  %0 = "tfl.reshape"(%arg0, %cst) : (tensor<13x21x?xf32>, tensor<2xi32>) -> tensor<*xf32>
-  func.return %0 : tensor<*xf32>
+// CHECK-LABEL: test_reshape_variable_shape
+// CHECK-SAME: %[[ARG0:.*]]: tensor<?xf32>
+// CHECK-SAME: %[[ARG1:.*]]: tensor<2xi32>
+// CHECK: %[[CONST_MINUS_1_TENSOR:.*]] = "tosa.const"() <{value = dense<-1> : tensor<2xi32>}> : () -> tensor<2xi32>
+// CHECK: %[[CONST_0_INDEX:.*]] = arith.constant 0 : index
+// CHECK: %[[CONST_1_TENSOR:.*]] = "tosa.const"() <{value = dense<1> : tensor<i32>}> : () -> tensor<i32>
+// CHECK: %[[CONST_0_I32:.*]] = arith.constant 0 : i32
+// CHECK: %[[SHAPE_PRODUCT_TENSOR:.*]] = linalg.reduce ins(%[[ARG1]] : tensor<2xi32>) outs(%[[CONST_1_TENSOR]] : tensor<i32>) dimensions = [0]
+// CHECK:   (%[[VAL_0:.*]]: i32, %[[VAL_1:.*]]: i32) {
+// CHECK:     %[[VAL_2:.*]] = arith.muli %[[VAL_0]], %[[VAL_1]] : i32
+// CHECK:     linalg.yield %[[VAL_2]] : i32
+// CHECK:   }
+// CHECK: %[[SHAPE_PRODUCT:.*]] = tensor.extract %[[SHAPE_PRODUCT_TENSOR]][] : tensor<i32>
+// CHECK: %[[SHAPE_PRODUCT_IS_NEGATIVE:.*]] = arith.cmpi slt, %[[SHAPE_PRODUCT]], %[[CONST_0_I32]] : i32
+// CHECK: %[[SHAPE_RESOLVED:.*]] = scf.if %[[SHAPE_PRODUCT_IS_NEGATIVE]] -> (tensor<2xi32>) {
+// CHECK:   %[[VAL_3:.*]] = tensor.dim %[[ARG0]], %[[CONST_0_INDEX]] : tensor<?xf32>
+// CHECK:   %[[VAL_4:.*]] = arith.index_cast %[[VAL_3]] : index to i32
+// CHECK:   %[[VAL_5:.*]] = math.absi %[[SHAPE_PRODUCT]] : i32
+// CHECK:   %[[VAL_6:.*]] = arith.divsi %[[VAL_4]], %[[VAL_5]] : i32
+// CHECK:   %[[VAL_7:.*]] = tensor.splat %[[VAL_6]] : tensor<2xi32>
+// CHECK:   %[[VAL_8:.*]] = arith.cmpi eq, %[[ARG1]], %[[CONST_MINUS_1_TENSOR]] : tensor<2xi32>
+// CHECK:   %[[VAL_9:.*]] = arith.select %[[VAL_8]], %[[VAL_7]], %[[ARG1]] : tensor<2xi1>, tensor<2xi32>
+// CHECK:   scf.yield %[[VAL_9]] : tensor<2xi32>
+// CHECK: } else {
+// CHECK:   scf.yield %[[ARG1]] : tensor<2xi32>
+// CHECK: }
+// CHECK: %[[RESULT:.*]] = tensor.reshape %[[ARG0]](%[[SHAPE_RESOLVED]]) : (tensor<?xf32>, tensor<2xi32>) -> tensor<?x?xf32>
```

Comment:
This lowering does not contain a single TOSA operation and should be handled with a different pass / set of rewriters. This way passes can conditionally support this style of lowering.